ibis/common/scope.py

```diff
@@ -0,0 +1,107 @@
+""" Module for scope
+    Scope is a dictionary mapping :class:`~ibis.expr.operations.Node`
+    subclass instances to concrete data, and the time context associate
+    with it(if any).
+    Scope is used in many backends to cache data for calculated ops.
+
+    Set scope kv pair: before setting the value op in scope we need to perform
+    the following check first:
+
+    Test if op is in scope yet
+    - No, then put op in scope, set timecontext to be the current timecontext
+    (None if timecontext is not present), set value to be the DataFrame or
+    Series of the actual data.
+    - Yes, then get the timecontext stored in scope for op as old_timecontext,
+    and compare it with current timecontext:
+    If current timecontext is a subset of old timecontext, that means we
+    already cached a larger range of data. Do nothing and we will trim data in
+    later execution process.
+    If current timecontext is a superset of old timecontext, that means we
+    need to update cache. Set value to be the current data and set timecontext
+    to be the current timecontext for op.
+    If current timecontext is neither a subset nor a superset of old
+    timcontext, but they overlap, or not overlap at all. For example this will
+    happen when there is a window that looks forward, over a window that looks
+    back. So in this case, we should not trust the data stored either, and go
+    on to execute this node. For simplicity, we update cache in this case as
+    well.
+
+"""
+from typing import Optional
+
+from ibis.expr.timecontext import TimeContextRelation, compare_timecontext
+from ibis.expr.typing import TimeContext
+
+
+def make_scope_item(op, result, timecontext: Optional[TimeContext]):
```

Comment:
I think it's confusing to me what's an scope item:

```
{op: {'value': result, 'timecontext': timecontext}}
```
This looks like a "scope"

From the doc of `get_scope_item`, "scope item" is just "result" here?

Reply:
I see, we may need some renaming here. `scope_item` is intended to hide the implementation of key-value inside.
```
op: {'value': result, 'timecontext': timecontext}
```
This key-value pair is an entry in `scope`.  And `get_scope_item` probably just get result out of this entry since that is what `get_scope` without timecontext usually does, although there is no such a `get_scope` method before.